cpp/benchmarks/groupby/group_max_multistream.cpp

```diff
@@ -0,0 +1,103 @@
+
+#include <nvbench/nvbench.cuh>
+
+template <typename Type>
+void groupby_max_multistream_helper(nvbench::state& state,
+                                    cudf::size_type num_rows,
+                                    cudf::size_type cardinality,
+                                    double null_probability)
+{
+  auto const keys = [&] {
+    data_profile const profile =
+      data_profile_builder()
+        .cardinality(cardinality)
+        .no_validity()
+        .distribution(cudf::type_to_id<int32_t>(), distribution_id::UNIFORM, 0, num_rows);
+    return create_random_column(cudf::type_to_id<int32_t>(), row_count{num_rows}, profile);
+  }();
+
+  auto const vals = [&] {
+    auto builder = data_profile_builder().cardinality(0).distribution(
+      cudf::type_to_id<Type>(), distribution_id::UNIFORM, 0, num_rows);
+    if (null_probability > 0) {
+      builder.null_probability(null_probability);
+    } else {
+      builder.no_validity();
+    }
+    return create_random_column(
+      cudf::type_to_id<Type>(), row_count{num_rows}, data_profile{builder});
+  }();
+
+  auto keys_view = keys->view();
+  auto gb_obj    = cudf::groupby::groupby(cudf::table_view({keys_view, keys_view, keys_view}));
+
+  auto const num_streams = state.get_int64("num_streams");
+
+  auto streams = cudf::detail::fork_streams(cudf::get_default_stream(), num_streams);
+
+  std::vector<std::vector<cudf::groupby::aggregation_request>> requests(num_streams);
+  for (int64_t i = 0; i < num_streams; i++) {
+    requests[i].emplace_back(cudf::groupby::aggregation_request());
+    requests[i][0].values = vals->view();
+    requests[i][0].aggregations.push_back(cudf::make_max_aggregation<cudf::groupby_aggregation>());
+  }
+
+  auto const mem_stats_logger = cudf::memory_stats_logger();
+  state.exec(
+    nvbench::exec_tag::sync | nvbench::exec_tag::timer, [&](nvbench::launch& launch, auto& timer) {
+      auto perform_agg = [&](int index) { gb_obj.aggregate(requests[index], streams[index]); };
+      timer.start();
+      for (int i = 0; i < num_streams; i++) {
+        perform_agg(i);
+      }
+      cudf::detail::join_streams(streams, cudf::get_default_stream());
+      timer.stop();
+    });
+
+  auto const elapsed_time = state.get_summary("nv/cold/time/gpu/mean").get_float64("value");
+  state.add_element_count(static_cast<double>(num_rows) / elapsed_time / 1'000'000., "Mrows/s");
+  state.add_buffer_size(
+    mem_stats_logger.peak_memory_usage(), "peak_memory_usage", "peak_memory_usage");
+}
+
+template <typename Type>
+void bench_groupby_max_multistream(nvbench::state& state, nvbench::type_list<Type>)
+{
+  auto const cardinality      = static_cast<cudf::size_type>(state.get_int64("cardinality"));
+  auto const num_rows         = static_cast<cudf::size_type>(state.get_int64("num_rows"));
+  auto const null_probability = state.get_float64("null_probability");
+
+  groupby_max_multistream_helper<Type>(state, num_rows, cardinality, null_probability);
+}
```

Comment:
Could we move the content of `groupby_max_multistream_helper` into this function?